File: 17_Letter_Combinations_of_a_Phone_Number.java
class Solution {

    private String[] letters = {"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};

    public List<String> letterCombinations(String digits) {
        ArrayList<String> result = new ArrayList<>();

        /* Sanity check */
        if(digits.equals(""))
            return result;

        StringBuilder currentString = new StringBuilder();
        dfs(result, digits, currentString, 0);
        return result;
    }

    private void dfs(List<String> result, final String digits, StringBuilder currentString, int depth) {

        /* Reached end, append to result array */
        if(currentString.length() == digits.length()) {
            result.add(currentString.toString());
        }

        /* Otherwise, dfs next level */
        else {
            /* Get index into letters array */
            int index = Character.getNumericValue(digits.charAt(depth)) - 2;

            for(int i = 0; i < letters[index].length(); ++i) {
                currentString.append(letters[index].charAt(i));
                dfs(result, digits, currentString, depth + 1);
                currentString.deleteCharAt(currentString.length() - 1);
            }
        }
    }

}

File: 17_Letter_Combinations_of_a_Phone_Number.cpp
class Solution {
 private:
  vector<string> data{"",    "",    "abc",  "def", "ghi",
                      "jkl", "mno", "pqrs", "tuv", "wxyz"};
  vector<string> dfs(const string& str) {
    vector<string> result;
    if (str.empty()) return result;
    if (str.length() == 1) {
      string temp = data[str[0] - '0'];
      for (int i = 0; i < temp.length(); ++i) {
        result.push_back(temp.substr(i, 1));
      }
    } else {
      vector<string> sub = dfs(str.substr(1, str.length() - 1));
      string first = data[str[0] - '0'];
      for (int i = 0; i < first.length(); ++i) {
        for (int j = 0; j < sub.size(); ++j) {
          result.push_back(first.substr(i, 1) + sub[j]);
        }
      }
    }
    return result;
  }

 public:
  vector<string> letterCombinations(string digits) { return dfs(digits); }
};